compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -159,6 +160,9 @@ trait TypeAssigner {
       TryDynamicCallType
     else if (qualType.isErroneous || name.toTermName == nme.ERROR)
       UnspecifiedErrorType
+    else if !isFullyDefined(qualType, ForceDegree.none) && isFullyDefined(qualType, ForceDegree.failBottom) then
+      // try again with more defined qualifier type
+      selectionType(tree, qual1)
```

Comment:
`isFullyDefined` traverses the whole type to instantiate type variables, but to be able to type a selection we really only need to make sure the top-level part of the type is instantiated, I've been experimenting with this separately since it would become even more necessary if we allow lambda binders to not be fully instantiated, if you want I can try to extract that into its own PR.

Reply:
Yes, please go ahead with this!

Reply:
I added a new method `couldInstantiateTypeVar` to `Inferencing` that does this.